src/DotVVM.Framework/Controls/CheckBox.cs

```diff
@@ -23,7 +24,17 @@ public IEnumerable CheckedItems
         }
         public static readonly DotvvmProperty CheckedItemsProperty =
             DotvvmProperty.Register<IEnumerable, CheckBox>(t => t.CheckedItems, null);
-        
+
+        protected internal override void OnLoad(IDotvvmRequestContext context)
+        {
+            if (CheckedItems != null && CheckedValue != null && 
+                CheckedItems.GetType().GetGenericArguments()[0] != CheckedValue.GetType())
+            {
+                throw new DotvvmControlException(this,
+                    $"Type of items in CheckedItems \'{CheckedItems.GetType().GetGenericArguments()[0].FullName}\' must be same as CheckedValue type \'{CheckedValue.GetType().FullName}\'.");
+            }
+            base.OnLoad(context);
```

Comment:
I'm sorry, but bindings have a `ResultType` property that is determined compile time and it should be used to determine if the value types are matching - this will not work in client-side templates, in case the value is null for another reason and will cause unnecessary binding evaluation (5 of them actually).

Another problem is that checking only the generic arguments is not correct at all - it will break on pure arrays, types implementing only `System.Collections.IEnumerable` and types that implement `IEnumerable<T>` but do not have type parameter itself. It is also possible that the type parameter means something completely different. Anyway, we have a `ReflectionUtils.GetEnumerableType` helper method that does exactly what you need here.